File: c/solutions.c
/*
Write a program that outputs all possibilities to put + or - or nothing
between the numbers 1, 2, ..., 9 (in this order) such that the result is
always 100. For example: 1 + 2 + 34 – 5 + 67 – 8 + 9 = 100.
*/
#include <stdio.h>
#include <stdlib.h>

static const int NUMBERS_SIZE = 9;
static const int NUMBERS[] = {1,2,3,4,5,6,7,8,9};
static const int SUM = 100;
static const int BUFFER_SIZE = 10;

void dump(int* buffer, int size)
{
    for (int i = 0; i < size; ++i)
    {
        if (i == 0)
            printf("%d", buffer[i]);
        else
            printf(" %s %d", (buffer[i] >= 0 ? "+" : "-"), abs(buffer[i]));
    }
    printf(" = %d\n", SUM);
}

void recurse(int* buffer, int idx, int size)
{
    // Check if we have recursed to depth
    if (idx >= NUMBERS_SIZE)
    {
        int result = 0;
        for (int i = 0; i < size; ++i)
            result += buffer[i];
        if (result == SUM)
            dump(buffer, size);
        return;
    }

    // Branch 1: Add number to previous
    buffer[size] = NUMBERS[idx];
    recurse(buffer, idx+1, size+1);

    // Branch 2: Subtract number from previous
    buffer[size] = -1 * NUMBERS[idx];
    recurse(buffer, idx+1, size+1);

    // Branch 3: Concatenate number to previous
    int prev = buffer[size-1];
    buffer[size-1] = (10 * abs(prev) + NUMBERS[idx]) * (prev >= 0 ? 1 : -1);
    recurse(buffer, idx+1, size);
    buffer[size-1] = prev;
}

int main(int argc, char* argv[])
{
    int buffer[BUFFER_SIZE];
    buffer[0] = NUMBERS[0];
    recurse((int*)buffer, 1, 1);
    return 0;
}

File: go/solutions.go
// Write a program that outputs all possibilities to put + or - or nothing
// between the numbers 1, 2, ..., 9 (in this order) such that the result is
// always 100. For example: 1 + 2 + 34 – 5 + 67 – 8 + 9 = 100.
package main

import (
    "fmt"
)

var NUMBERS [9]int = [9]int{1,2,3,4,5,6,7,8,9}
var SUM int = 100

func print(solution []int) {
    for i, v := range solution {
        switch {
        case i == 0:
            fmt.Printf("%d ", v)
        case v < 0:
            fmt.Printf("- %d ", -1 * v)
        default:
            fmt.Printf("+ %d ", v)
        }
    }
    fmt.Printf(" = %d\n", SUM)
}

func concatenate(lhs, rhs int) (result int) {
    sign := 1
    val := lhs
    if lhs < 0 {
        sign = -1
        val = -1 * lhs
    }
    result = sign * (10 * val + rhs)
    return
}

func recurse(buf []int, idx int) {
    // Check if we have recursed to depth
    if idx >= len(NUMBERS) {
        result := 0
        for _, v := range buf {
            result += v
        }
        if result == SUM {
            print(buf)
        }
        return
    }

    // Branch 1: Add number to previous
    buf = append(buf, NUMBERS[idx])
    recurse(buf, idx+1)
    buf = buf[:len(buf)-1]

    // Branch 2: Subtract number from previous
    buf = append(buf, -1 * NUMBERS[idx])
    recurse(buf, idx+1)
    buf = buf[:len(buf)-1]

    // Branch 3: Concatenate number to previous
    prev := buf[len(buf)-1]
    buf[len(buf)-1] = concatenate(prev, NUMBERS[idx])
    recurse(buf, idx+1);
    buf[len(buf)-1] = prev;
}

func main() {
    buf := make([]int, 0, 9)
    buf = append(buf, NUMBERS[0])
    recurse(buf, 1)
}
